### DISCUSSION TITLE
Handle google oauth in front end (react-admin) and back with fastapi-users (with oauth on back end)

### DISCUSSION THREAD

**Discussion by diegocgaona** [fastapi-users/fastapi-users#1081] [Q&A]
I have some doubts how to handle the front/back flow using FastApi (with FastApi-Users) and React-Admin on front end. (all in their latest versions) I'm using this part of the documentation as reference: [FastApi-Users Oauth with Beanie](https://fastapi-users.github.io/fastapi-users/10.1/configuration/oauth/#beanie) (and the backend code used for this is almost the same from the full example)

I want to make the auth on the back end and only return the access_token and token_type to the front (for example, the front will call for example /auth/google and receive the response after the successful login). Like:

```
{
    "access_token": "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ1c2VyX2lkIjoiOTIyMWZmYzktNjQwZi00MzcyLTg2ZDMtY2U2NDJjYmE1NjAzIiwiYXVkIjoiZmFzdGFwaS11c2VyczphdXRoIiwiZXhwIjoxNTcxNTA0MTkzfQ.M10bjOe45I5Ncu_uXvOmVV8QxnL-nZfcH96U90JaocI",
    "token_type": "bearer"
}
```
I want to put a button in the front which would go to the back (in an endpoint like /auth/google for example), in the back the user will be redirected (open a new tab I think) to the authorization_url received from /auth/google/authorize (endpoint in my back end, default from FastApi-Users which returns only the authorization_url, will make the login with Google, then it would give the response back to the front with the access_token.

What I'm struggling is: If I call from my front the GET /authorize, I will receive the authorization_url, but then if I go to the url by the front, is a Google url, which will answer to my GET /callback in the back (and save to the DB), and my front will not receive the access_token.

It is possible to create an endpoint or modify the GET /authorize, to in the back redirect to the authorization_url, the it will call the GET /callback, record the data to the DB, and then return the response to the GET /auth/google (or the endpoint GET /authorize for example) so the front will receive the access_token and token_type?

This way, my front will not receive anything from Google, only the token for my backend which I would prefer.

Thanks in advance

**Comment by frankie567** [comment]
What I usually do in those cases is to make the redirect URL pointing to an URL on my **frontend**. This way, I can have JS code catch the authorization code in the query parameters and make an XHR request to my `/callback` API. Then, the frontend directly gets the response of the callback and can use/store the token.

**Reply by yibi** [reply]
Does this mean hardcoding of redirect_url when calling get_oauth_router?

Can we pass redirect_url when calling /authorize?

**Reply by zinoubm** [reply]
any news on that?

**Comment by diegocgaona** [comment]
Thanks @frankie567. I will try that.
If you have an example of the JS code catch to make the request to `/callback` I would appreciate.

And the possibility to do the opposite, the backend send the parameters to the front? It have a way to do it?

**Reply by zinoubm** [reply]
Did you solve the problem?

**Reply by diegocgaona** [reply]
Hi @zinoubm, see the https://github.com/fastapi-users/fastapi-users/discussions/1111
If I remember correctly, I used an extra cookie (and a global variable in Python) to know that the login was made by oauth and then redirect to my front url (are the last lines of the code in the referenced issue):

```
response.status_code = status.HTTP_302_FOUND
        response.headers["Location"] = f"{FRONT_END_URL}{FRONT_HOME_PATH}"
```

And more info on: https://fastapi.tiangolo.com/advanced/response-headers/

Hope it helps!
